src/test/java/com/aws/greengrass/componentmanager/builtins/GreengrassRepositoryDownloaderTest.java

```diff
@@ -169,6 +169,55 @@ void GIVEN_http_connection_error_WHEN_attempt_download_THEN_retry_called(Extensi
         assertThat(e.getLocalizedMessage(), containsStringIgnoringCase("Failed to download artifact"));
     }
 
+    @Test
+    void GIVEN_get_download_size_response_as_5xx_WHEN_attempt_get_download_size_THEN_retry(ExtensionContext context) throws Exception {
+        ignoreExceptionOfType(context, RetryableServerErrorException.class);
+        // build downloader
+        Path mockArtifactPath = ComponentTestResourceHelper
+                .getPathForTestPackage(ComponentTestResourceHelper.MONITORING_SERVICE_PACKAGE_NAME, "1.0.0")
+                .resolve("monitor_artifact_100.txt");
+        String checksum = Base64.getEncoder()
+                .encodeToString(MessageDigest.getInstance(SHA256).digest(Files.readAllBytes(mockArtifactPath)));
+        ComponentArtifact artifact = ComponentArtifact.builder().algorithm(SHA256).checksum(checksum)
+                .artifactUri(new URI("greengrass:774pP05xtua0RCcwj9uALSdAqGr_vC631EdOBkJxnec=/artifact.txt")).build();
+        ComponentIdentifier pkgId = new ComponentIdentifier("CoolService", new Semver("1.0.0"));
+
+        lenient().when(componentStore.getRecipeMetadata(pkgId)).thenReturn(new RecipeMetadata(TEST_ARN));
+
+        Path testCache = ComponentTestResourceHelper.getPathForLocalTestCache();
+        Path saveToPath = testCache.resolve("CoolService").resolve("1.0.0");
+        Files.createDirectories(saveToPath);
+
+        GreengrassRepositoryDownloader downloader =
+                spy(new GreengrassRepositoryDownloader(clientFactory, pkgId, artifact, saveToPath, componentStore));
+
+        assertThat(downloader.getArtifactFilename(), is("artifact.txt"));
+
+        // mock requests to get downloadSize and local file name
+        GetComponentVersionArtifactResponse result =
+                GetComponentVersionArtifactResponse.builder()
+                        .preSignedUrl("https://www.amazon.com/artifact.txt").build();
+        when(client.getComponentVersionArtifact(getComponentVersionArtifactRequestArgumentCaptor.capture()))
+                .thenReturn(result);
+
+        // mock requests to return 500 error code
+        doReturn(httpClient).when(downloader).getSdkHttpClient();
+        doReturn(request).when(httpClient).prepareRequest(any());
+        when(request.call())
+                .thenReturn(HttpExecuteResponse.builder()
+                        .response(SdkHttpResponse.builder().statusCode(500).build())
+                        .build())
+                .thenReturn(HttpExecuteResponse.builder()
+                        .response(SdkHttpResponse.builder().statusCode(HTTP_OK)
+                                .putHeader(CONTENT_LENGTH_HEADER, String.valueOf(Files.size(mockArtifactPath))).build())
+                        .responseBody(AbortableInputStream.create(Files.newInputStream(mockArtifactPath)))
+                        .build());
+
+        downloader.getDownloadSize();
+        assertNotEquals(Files.size(mockArtifactPath), -1);
```

Comment:
Can you verify that it retried by like other tests?

Reply:
For this test, the method which invoked returns a status code in the original function instead of catching an exception. The test logic is: return the response code to be 500, thenReturn a valid code, assert the file size is valid.

Reply:
Same here, we need to verify the number of retry attempts, let's try to find a way to do that like your other tests, verifying the response that you eventually get after the last successful attempt is secondary. Just like other tests